torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -625,6 +698,12 @@ void Reducer::finalize_backward() {
       finalize_bucket_dense(bucket);
     }
   }
+
+  // Reset unused parameter accounting.
+  for (auto& local_used : local_used_maps_) {
+    local_used.fill_(0);
+  }
+  local_used_maps_reduced_ = false;
```

Comment:
The work object needs to be cleaned up as well, but like I said in the previous comment, it's probably best to do this lazily if it wasn't needed in the first place. Then the only overhead we end up paying is the H2D memcpy and the kernel launch cost.

Reply:
Not sure if I'm following ... We already do lazy wait as you suggested previously. For example, if all params are used locally, we just skip work wait and D2H copy. For such case, the overhead is just H2D copy and the allreduce kernel launch cost.

Reply:
The difference with the bucket work is that we wait on all of them.

If we don't wait, it is possible for the next iteration to interfere and write to the device-side memory for `local_unused_maps_dev_` while the allreduce of the previous iteration is still going. Sure, it's a long shot, but we must eliminate the possibility of this happening at all.

Reply:
I see. That's valid point. I'll address. Thanks @pietern .

Reply:
Done with bc1c493.